sbin/anti-evil-maid-lib-tpm2

```diff
@@ -0,0 +1,319 @@
+    _opw=$(cat "$TPM_OWNER_PASSWORD_FILE")
+    # create a write-once and read-by-anyone NVRAM area
+    _session="$(mktemp)"
+    _policy="$(mktemp)"
+    tpm2_startauthsession -S "$_session"
+    tpm2_policycommandcode -Q -S "$_session" -L "$_policy" TPM2_CC_NV_Read
+    tpm2_flushcontext "$_session"
+    rm "$_session"
+    tpm2_nvdefine -Q -s 20 -P "$_opw" -L "$_policy" "$_tpm_id_index" \
+                  -a "policyread|writedefine|writeall|ownerwrite|ownerread"
+    rm "$_policy"
+    # generate a random ID and write it into NVRAM
+    head -c 20 /dev/random |
+      tpm2_nvwrite -Q -C o -P "$_opw" -i - "$_tpm_id_index"
+    # lock the area to prevent changing the ID even by the owner
+    tpm2_nvwritelock -C o -P "$_opw" "$_tpm_id_index"
+}
+
+postprovisioning() {
+    # takes care of updating /var/lib/tpms after a successful provisioning by
+    # provisiontpmid
+    _tpmid=$(tpm2_id)
+    mkdir -p "/var/lib/tpms/$_tpmid"
+}
+
+checksrkpass() {
+    # checks whether contents of $SRK_PASSWORD_CACHE file is a valid SRK
+    # password, signals result with exit code
+
+    # `echo` is needed because empty input doesn't work and `echo` it provides
+    # '\n'
+    echo | tpm2_create -Q -C "$TPM2_SRK_HANDLE" -i - \
+                       -P "str:$(cat "$SRK_PASSWORD_CACHE")"
+}
+
+tpmpcrextend() {
+    # extends a PCR with a hash value of a suitable type
+    _pcr=$1
+    _hash=$2
+    tpm2_pcrextend "$_pcr:sha256=$_hash"
+}
+
+tpmsealprepare() {
+    # does necessary preparations before the use of tpmsealdata, accepts path
+    # to media-specific storage of sealed data
+    _dir=$1
+
+    # this recreates a sealing key on every run to pick up configuration/PCR
+    # changes if there were any
+
+    _pcrs=$(printf %s "$SEAL" | grep -Eo '\b[0-9]+\b')
+    _pcrs=sha256:${_pcrs//$'\n'/,}
+
+    _policy="$(mktemp)"
+    _session="$(mktemp)"
+
+    # make a suitable PCR policy
+    tpm2_startauthsession -S "$_session"
+    tpm2_policypcr -Q -S "$_session" -l "$_pcrs" -L "$_policy"
+    tpm2_flushcontext "$_session"
+
+    # make a key for sealing
+    tpm2_create -Q -C "$TPM2_SRK_HANDLE" -Gaes128 \
+                -P "str:$(cat "$SRK_PASSWORD_CACHE")" -L "$_policy" \
+                -u "$_dir/key.pub" -r "$_dir/key.priv"
+    echo "$_pcrs" > "$_dir/key.pcrs"
+
+    rm "$_session"
+    rm "$_policy"
+}
+
+tpmsealdata() {
+    # seals source specified by second argument into destination specified by
+    # the third one, non-empty first argument signifies empty SRK password, the
+    # forth argument specifies path to AEM media-specific storage
+    _infile=$2
+    _outfile=$3
+    _dir=$4
+
+    _ctx="$(mktemp)"
+    tpm2_load -Q -C "$TPM2_SRK_HANDLE" -P "str:$(cat "$SRK_PASSWORD_CACHE")" \
+              --private "$_dir/key.priv" --public "$_dir/key.pub" \
+              -c "$_ctx" || return 1
+
+    tpm2_encryptdecrypt -Q -c "$_ctx" -p "pcr:$_pcrs" \
+                        -o "$_outfile" "$_infile" || return 1
```

Comment:
This fails for me, full log: https://gist.github.com/marmarek/98e305332ac828e2562cd7f125f7fa31
It could be something wrong with my TPM, I've never used it in this system, but earlier steps succeeded, so I'd expect it to work fine...

Reply:
I guess this won't print anything for you, then:
```
sudo tpm2_getcap commands | grep -i Encryptdecrypt
```

I didn't know it's optional, but https://github.com/tpm2-software/tpm2-tools/issues/3180#issuecomment-1354996963 says:
> You're TPM doesn't support EncryptDecrypt, **most don't.**

Don't know if `openssl` is already in initrd, but something capable of decryption will be needed there.

Reply:
> I guess this won't print anything for you, then:

Indeed, no EncryptDecrypt listed. I have only `TPM2_CC_RSA_Encrypt`

> Don't know if `openssl` is already in initrd, but something capable of decryption will be needed there.

The failure I see is during sealing, which is outside of initrd already. But the same issue will apply to unsealing, right? Adding `openssl` or similar to initrd should be okay, but please check how much more space it will need (together with necessary deps).

Reply:
> But the same issue will apply to unsealing, right?

Yes, same command-line and TPM command handles this (sealing/unsealing of data stored outside of TPM2).

> check how much more space it will need (together with necessary deps).

OK, will post before/after sizes.  I see `libssl` already there, so it shouldn't be much.

Reply:
Updated to use `openssl enc`. `initramfs-5.15.94-1.qubes.fc37.x86_64.img` grows from 34.47 MiB to 34.83 MiB.